File: backend/app/db/models/risk.py
"""Foresight AI - Risk State & Attack Path Transition Database Models."""

from datetime import datetime, timezone
from typing import Optional
import uuid
from sqlalchemy import DateTime, Float, Index, Integer, JSON, String
from sqlalchemy.orm import Mapped, mapped_column
from app.db.base import Base, TimestampMixin


class RiskStateRecord(Base, TimestampMixin):
    """Persistent historical snapshot of perimeter risk state evaluations with anti-flutter metadata."""
    __tablename__ = "risk_state_records"

    id: Mapped[str] = mapped_column(
        String(36),
        primary_key=True,
        default=lambda: str(uuid.uuid4()),
        index=True
    )
    evaluation_id: Mapped[str] = mapped_column(String(64), index=True, nullable=False)
    timestamp: Mapped[datetime] = mapped_column(DateTime(timezone=True), index=True, nullable=False)
    current_state: Mapped[str] = mapped_column(String(32), index=True, nullable=False)  # NORMAL, WATCH, SUSPICIOUS, ELEVATED, CRITICAL
    previous_state: Mapped[str] = mapped_column(String(32), nullable=False)
    max_calibrated_probability: Mapped[float] = mapped_column(Float, default=0.0, nullable=False)
    triggering_horizon_minutes: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    active_alert_horizons: Mapped[list] = mapped_column(JSON, default=list, nullable=False)
    anomaly_score: Mapped[float] = mapped_column(Float, default=0.0, nullable=False)
    uncertainty_level: Mapped[str] = mapped_column(String(32), default="LOW", nullable=False)
    evidence_summary: Mapped[list] = mapped_column(JSON, default=list, nullable=False)
    top_risk_features: Mapped[list] = mapped_column(JSON, default=list, nullable=False)
    attack_path_summary: Mapped[dict] = mapped_column(JSON, default=dict, nullable=False)
    is_hysteresis_dampened: Mapped[bool] = mapped_column(default=False, nullable=False)
    model_version: Mapped[str] = mapped_column(String(50), nullable=False)

    __table_args__ = (
        Index("idx_risk_state_time", "current_state", "timestamp"),
    )


class RiskTransitionRecord(Base, TimestampMixin):
    """Auditable state transition event logging security escalation / de-escalation."""
    __tablename__ = "risk_transition_records"

    id: Mapped[str] = mapped_column(
        String(36),
        primary_key=True,
        default=lambda: str(uuid.uuid4()),
        index=True
    )
    transition_id: Mapped[str] = mapped_column(String(64), index=True, nullable=False)
    timestamp: Mapped[datetime] = mapped_column(DateTime(timezone=True), index=True, nullable=False)
    previous_state: Mapped[str] = mapped_column(String(32), nullable=False)
    new_state: Mapped[str] = mapped_column(String(32), index=True, nullable=False)
    forecast_probability: Mapped[float] = mapped_column(Float, nullable=False)
    anomaly_score: Mapped[float] = mapped_column(Float, nullable=False)
    uncertainty: Mapped[str] = mapped_column(String(32), default="LOW", nullable=False)
    triggering_horizon: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    reason: Mapped[str] = mapped_column(String(500), nullable=False)
    triggering_evidence: Mapped[list] = mapped_column(JSON, default=list, nullable=False)
    model_version: Mapped[str] = mapped_column(String(50), nullable=False)

    __table_args__ = (
        Index("idx_transition_time_state", "new_state", "timestamp"),
    )

File: frontend/src/lib/api/risk.ts
/**
 * Foresight AI - Attack Path & Risk State Engine API Service
 */

import { apiClient } from './client';

export type RiskState = 'NORMAL' | 'WATCH' | 'SUSPICIOUS' | 'ELEVATED' | 'CRITICAL';

export interface AttackPathNode {
  stage: string;
  is_current: boolean;
  is_predicted: boolean;
  is_forecasted: boolean;
  probability?: number;
  evidence_strength: 'LOW' | 'MEDIUM' | 'HIGH' | 'VERIFIED' | 'INSUFFICIENT_EVIDENCE' | 'GRAPH_KNOWLEDGE_ONLY';
  status: 'OBSERVED' | 'PREDICTED' | 'GRAPH_POSSIBLE' | 'POTENTIAL_UNVALIDATED' | 'INSUFFICIENT_EVIDENCE';
  supporting_features: Array<Record<string, any>>;
  timestamp?: string;
}

export interface AttackPathEdge {
  source_stage: string;
  target_stage: string;
  transition_probability?: number;
  is_forecasted: boolean;
  transition_horizon_minutes?: number;
  evidence_reasons: string[];
  uncertainty: string;
  status: string;
}

export interface AttackPathForecast {
  forecast_id: string;
  forecast_timestamp: string;
  target_entity: string;
  current_stage?: string;
  predicted_next_stage?: string;
  subsequent_stages: string[];
  probability?: number;
  uncertainty: string;
  evidence: string[];
  supporting_features: Array<Record<string, any>>;
  horizon_minutes?: number;
  target_timestamp?: string;
  transition_confidence_status: string;
  model_version: string;
  graph_nodes: AttackPathNode[];
  graph_edges: AttackPathEdge[];
}

export interface RiskStateTransition {
  transition_id: string;
  timestamp: string;
  previous_state: RiskState;
  new_state: RiskState;
  triggering_evidence: string[];
  forecast_probability: number;
  anomaly_score: number;
  uncertainty: string;
  triggering_horizon?: number;
  reason: string;
  model_version: string;
}

export interface RiskStateEvaluation {
  evaluation_id: string;
  timestamp: string;
  current_state: RiskState;
  previous_state: RiskState;
  state_duration_seconds: number;
  cycles_in_state: number;
  is_hysteresis_dampened: boolean;
  hysteresis_notes: string[];
  max_calibrated_probability: number;
  triggering_horizon_minutes?: number;
  active_alert_horizons: number[];
  anomaly_score: number;
  conformal_coverage_status: string;
  uncertainty_level: string;
  evidence_summary: string[];
  top_risk_features: Array<Record<string, any>>;
  attack_path_summary?: AttackPathForecast;
  model_version: string;
}

export interface RiskTimelineResponse {
  current_evaluation: RiskStateEvaluation;
  timeline_transitions: RiskStateTransition[];
  historical_evaluations: RiskStateEvaluation[];
}

export const riskApi = {
  async getCurrentState(targetEntity = 'GLOBAL_PERIMETER'): Promise<RiskStateEvaluation> {
    return apiClient<RiskStateEvaluation>(`/risk/state?target_entity=${targetEntity}`);
  },

  async getTimeline(targetEntity = 'GLOBAL_PERIMETER', limit = 20): Promise<RiskTimelineResponse> {
    return apiClient<RiskTimelineResponse>(`/risk/timeline?target_entity=${targetEntity}&limit=${limit}`);
  },

  async getAttackPath(targetEntity = 'GLOBAL_PERIMETER'): Promise<AttackPathForecast> {
    return apiClient<AttackPathForecast>(`/risk/attack-path?target_entity=${targetEntity}`);
  },

  async getTransitions(limit = 20): Promise<RiskStateTransition[]> {
    return apiClient<RiskStateTransition[]>(`/risk/transitions?limit=${limit}`);
  },
};
